.github/workflows/ci.yaml

```diff
@@ -9,27 +9,46 @@ on:
 jobs:
 
   build-test:
-    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        os: [ubuntu-latest, macos-latest, windows-latest]
+      fail-fast: false
+    runs-on: ${{ matrix.os }}
     steps:
     - uses: actions/checkout@v4
     - name: Set up Go
       uses: actions/setup-go@v5
       with:
         go-version-file: 'go.mod'
         cache: true
+    - name: Cache Go modules
+      uses: actions/cache@v3
+      with:
+        path: |
+          ~/.cache/go-build
+          ~/go/pkg/mod
+        key: ${{ runner.os }}-go-${{ hashFiles('go.sum') }}
+        restore-keys: |
+          ${{ runner.os }}-go-
     - name: build 
       run: make build
 
   unit-test:
```

Comment:
Our go-test does not depend on a working Kubernetes cluster.

So we might split this `unit-test` into two, `unit-test` and probably `e2e-test`:

- `unit-test` should not depend on K8s, and simply setting up Go, go cache, and running `go test ./...` (via make) should do the job.
- `e2e-test` should run the e2e tests written in shellscript, which depends on a working Kubernetes cluster.

This way, we won't need a lot of `if: ${{ matrix.os == 'ubuntu-latest' }}`, because:

- `unit-test` should run all the steps against all the supported oses.
- `e2e-test` should run all the steps against all the supported oses. (Not sure if `kind` is able to run on all the managed GHA runners, especially macos and windows, but that's the idea.)

WDYT?

Reply:
that sounds like an idea that can work: One question I have here is, which tests? are unit and which are e2e they are not listed from what it seems